File: 0117-populating-next-right-pointers-in-each-node-ii/0117-populating-next-right-pointers-in-each-node-ii.rb
# Definition for Node.
# class Node
#   attr_accessor :val, :left, :right, :next
#   def initialize(val)
#     @val = val
#     @left, @right, @next = nil, nil, nil
#   end
# end

# @param {Node} root
# @return {Node}
def is_leaf(node)
  !node.left && !node.right
end

def next_child(node)
  !node ? nil : node.left ? node.left : node.right
end

def last_child(node)
  !node ? nil : node.right ? node.right : node.left
end

def connect(root)
  this_cur, this_nxt, nxt_lvl = root, nil, next_child(root)
  while this_cur && nxt_lvl
    this_cur.left.next = this_cur.right if this_cur.left && this_cur.right
    if last_child(this_cur) && this_nxt
      this_nxt = this_nxt.next while this_nxt && is_leaf(this_nxt)
      last_child(this_cur).next = next_child(this_nxt)
    end
    this_cur = this_nxt
    this_nxt = this_nxt.next if this_nxt
    if !this_cur
      this_cur = nxt_lvl
      this_nxt = this_cur
      this_nxt = this_nxt.next while this_nxt && is_leaf(this_nxt)
      nxt_lvl = !this_nxt || is_leaf(this_nxt) ? nil : next_child(this_nxt)
      this_nxt = this_cur.next
    end
  end
  root
end

File: 0117-populating-next-right-pointers-in-each-node-ii/0117-populating-next-right-pointers-in-each-node-ii.ts
/**
 * Definition for Node.
 * class Node {
 *   val: number
 *   left: Node | null
 *   right: Node | null
 *   next: Node | null
 *   constructor(val?: number, left?: Node, right?: Node, next?: Node) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 *     this.next = (next===undefined ? null : next)
 *   }
 * }
 */
function connect(root: Node | null): Node | null {
  function isLeaf(node: Node): boolean {
    return !node.left && !node.right
  }
  
  function nextChild(node: Node | null): Node | null {
    return !node ? null : (node.left ? node.left : node.right)
  }
  
  function lastChild(node: Node | null): Node | null {
    return !node ? null : (node.right ? node.right : node.left)
  }
  
  let thisCur: Node | null = root
  let thisNxt: Node | null
  let nxtLvl: Node | null = nextChild(thisCur)
  
  while (thisCur && nxtLvl) {
    if (thisCur.left && thisCur.right) thisCur.left.next = thisCur.right
    if (lastChild(thisCur) && thisNxt) {
      while (thisNxt && isLeaf(thisNxt)) thisNxt = thisNxt.next
      lastChild(thisCur).next = nextChild(thisNxt)
    }
    thisCur = thisNxt
    if (thisNxt) thisNxt = thisNxt.next
    if (!thisCur) {
      thisCur = nxtLvl
      thisNxt = thisCur
      while (thisNxt && isLeaf(thisNxt)) thisNxt = thisNxt.next
      nxtLvl = !thisNxt || isLeaf(thisNxt) ? null : nextChild(thisNxt)
      thisNxt = thisCur.next
    }
  }
  return root
}